### PR TITLE
Add `PluginUITexture` import option for images

### PR DESCRIPTION
**Pull request by RandomShaper**
When you write a plugin requiring some textures for its in-editor UI, you face the problem that they need to be imported. That's a problem especially if your plugin needs to be able to load them from the very beginning (because of preloads or because it needs to display some UI as soon as it's activated).

Normally, textures in Godot are imported so they are converted to proper formats depending on the import kind and settings, and that's good. However, in this scenario there's no ambiguity: you need to plug some textures into some `Control`s and you just need them to work.

To meet that need, this PR adds a new import mode for images: `PluginUITexture`. What it does is importing the image as an `ImageTexture` with sensible flags for UI use (filter and mip-maps). This mode will have the original image loaded instead of any import "artifact" that is normally generated by other modes. For that, this mode is flagged as dummy.

(The mode could have been named `ImageTexture`, but by giving it the name of `PluginUITexture` it's only reasonable use case is made explicit and confusion among the other import modes is avoided.)

Finally, it's only enabled in tools builds. If a texture imported this way is attempted to be loaded in an exported game, a no-loader-found-for-this-resource error will be thrown.

---
**This code is generously donated by IMVU.**

### CODE REVIEW

**Comment by reduz** [comment]
to be honest, Its probably better to just make sure textures are imported before the plugin is loaded.

**Comment by RandomShaper** [comment]
That may not be easy in every situation. For instance, if some script in the plugin is loaded as a singleton that preloads some scene needing a texture for UI.

I think this change adds a lot of benefit at a very little code cost. When you see it running and you can use textures without importing them it looks super clean. For plugins containing no assets, those that have only an editor side, this makes them plug & play.

**Comment by reduz** [comment]
This is why the editor has code to register plugins, you can put your singleton code in there. IMO this is a problem with a badly written plugin, more than a required feature. I am against adding this kind f hack.

**Comment by reduz** [comment]
I think its probably easier to do have a function that does a yield until the import process is complete, so you can do whathever you want there.

**Comment by reduz** [comment]
think about it, if we start adding this sort of hacks for this, then many other users will start asking for hacks f or other types of resources. Lets find a proper way to do it and document it.

**Comment by avencherus** [comment]
> IMO this is a problem with a badly written plugin

Maybe it's unrelated here, but as it is now, the first time a plugin loads it will fail to load textures.
https://github.com/godotengine/godot/issues/17483

It has to be restarted, because it seems like the import and the subsequent file it creates happens after the registration of the plugin.

I get this all the time when cloning my repositories.  The first run spews errors.  This is with a basic plugin that is simply using `add_custom_type()` and has a texture provided for an icon.

If a fix is to be had, I would also prefer it to just work without any special additional steps, like configuring dozens of texture imports with clicking and typing about the UI.

**Comment by RandomShaper** [comment]
Maybe in a future I'll come back with something different, but in the meantime I'm just closing this PR as not of enough general interest.

**Comment by follower** [comment]
FYI: Thanks to @avencherus pointing out https://github.com/godotengine/godot/issues/39313#issuecomment-643914353 which suggests a potential workaround for at least some cases (specifically using a texture atlas but maybe even image texture resource might be sufficient).

----

**TL;DR:** Editor plugins *shouldn't* attempt to use the resource system during initialization despite documentation stating otherwise.

Suggested workarounds:

 * Delay resource loading in an editor plugin until it's known that "first reimport" has occurred. (It's currently unclear when exactly that is--but it's some point after a project scene has been opened.)

 * Avoid using the resource system at all and rely on direct loading of assets via e.g. `ImageLoader` from a plugin subdirectory protected by `.gdignore`. (Only works for purely editor-only plugins and maybe `add_custom_type()` icons.) [To be confirmed.]

----

### Additional context

Some of this is already covered in my existing https://github.com/godotengine/godot/issues/17483#issuecomment-587568375 but here's some extra specifics...

The code [around this PR](https://github.com/godotengine/godot/pull/35082/files#diff-b6d74ccf775245adb707953658e4defd) is a useful pointer to the underlying issue. Specifically, the code here which enables the plugins...: https://github.com/godotengine/godot/blob/02f7908d460bea7d580fe7526949be7f294e6783/editor/editor_node.cpp#L404-L416

...occurs before this comment (which appears in a number of places): https://github.com/godotengine/godot/blob/02f7908d460bea7d580fe7526949be7f294e6783/editor/editor_node.cpp#L430

Thus any plugin initialization code *cannot* rely on the resource system or it will break on the first load of a project.

*(Aside: I vaguely seem to recall there's a related issue I noticed with first import/loads from the Project Manager or Asset Library templates due to the same issue?)*

### Waiting for resource reimport

The "waiting" _\*cough\*_ mentioned in the code comment above is related to the state of these...:

https://github.com/godotengine/godot/blob/02f7908d460bea7d580fe7526949be7f294e6783/editor/editor_node.h#L390-L392

...which affect the operation of the following methods (I don't remember the full details but I think there's a `xxxx_changed` signal (via a filesystem scan?) that is emitted which *then* triggers the reimport):

https://github.com/godotengine/godot/blob/7f6767470dda159832162e0670ecbe1fbdfa7e71/editor/editor_node.cpp#L675-L692

https://github.com/godotengine/godot/blob/7f6767470dda159832162e0670ecbe1fbdfa7e71/editor/editor_node.cpp#L645-L673

So, basically, if a plugin relies on the resource loading system (during plugin initialisation, I guess) it'll be broken on first load because the resource "reimport" (i.e. first import) won't have happened.

Given that [the current custom plugin documentation](https://docs.godotengine.org/en/3.2/tutorials/plugins/editor/making_plugins.html#a-custom-node) advises to use `preload()` this means anyone who follows it will have a broken first load. 

### When should editor plugins first attempt to use the resource system?
 
As I think on it, maybe part of the issue is that this advice in the documentation is incorrect when it comes to editor plugin initialization:

> It’s important to deal with initialization and clean-up of resources. A good practice is to use the virtual function _enter_tree() to initialize your plugin and _exit_tree() to clean it up. Thankfully, the dialog generates these callbacks for you.

If first reimport doesn't happen until *after* plugin initialization then `_enter_tree()` *isn't* where resource-related initialization should occurs. (I *think* because the "tree" in this case is the *editor* tree not a scene tree.)

Obviously the resources need to get reimported before e.g. a dialog is first displayed, so any resource-related initialization should occur in a hook that's called *after* reimport completes.

### A complication originating from per-project plugins?

Actually, that makes me think this is somewhat related to a potential design/implementation issue due to conflating the editor scene tree with the project scene tree--since addons are per-project not editor.

### An existing solution already applied elsewhere

While re-researching all this I discovered https://github.com/godotengine/godot/commit/9131f70d8a9a908b14aa660b15c0bde5c4a789e4 which (maybe unintentionally?) fixes the equivalent issue for custom class icons by changing from using `ResourceLoader`:

https://github.com/godotengine/godot/blob/4689ece7415be4a598e5a881d64b1201bf8ce0e0/editor/editor_node.cpp#L3694-L3697

to using `ImageLoader`: https://github.com/godotengine/godot/blob/9131f70d8a9a908b14aa660b15c0bde5c4a789e4/editor/editor_node.cpp#L3644

Which obviously bypasses the delayed first reimport entirely.

But `class_name` is [documented](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_basics.html#classes) to take a string with the icon path whereas `add_custom_type()` takes a texture which implies use of `load()` or `preload()` for example.

### An additional import related wrinkle...

I've also just realised that there's any additional issue which I mentioned in my earlier comment that tried to go down the `ImageLoader` route from GDScript:

> (In addition, in the above patch I'm removing a more complex workaround which used Image.load() & ImageTexture.create_from_image() which results in icons always loading correctly but also generates a yellow warning about accessing an image as a resource. I think the warning might be possible to be avoided by putting the image in a directory excluded with .gdignore but by that stage I'd shaved so many yaks that I stopped before testing that.)

### Inconclusion...

So, depending on where one sits, this could be an engine bug, a design bug or a documentation bug...or some combination of all three. :)

*{ This probably would've been better in https://github.com/godotengine/godot/issues/17483 so I'll at least put the TL;DR: there too. }*
